### DISCUSSION TITLE
didjvu promises to segment layers, ever tried it?

### DISCUSSION THREAD

**Discussion by rmast** [trufanov-nok/minidjvu-mod#21] [General]
https://wiki.ubuntuusers.de/didjvu/ 

It's German, but translated to English it states:

> Colored multi-layer DjVu files cannot be created with open source programs without detours, since the source code for segmentation has not been open sourced. Jakub Wilk's didjvu program is a fully-fledged alternative. It uses the open source project Gamera to split the foreground and background levels. With this decomposition, it is also easily possible for "private" users to extract a bitonal layer of an image template and to create a multi-layer DjVu file; this can then be provided with a text layer, e.g. with the ocrodjvu program.

The resulting file, with the original minidjvu, is already small:

[robert3.zip](https://github.com/trufanov-nok/minidjvu-mod/files/7384383/robert3.zip)

It doesn't contain a FGBZ but a FG44:

djvudump robert3.djvu
  FORM:DJVU [21106] 
    INFO [10]         DjVu 2480x3500, v24, 72 dpi, gamma=2.2
    Sjbz [11959]      JB2 bilevel data
    FG44 [6039]       IW4 data #1, 100 slices, v1.2 (color), 414x584
    BG44 [1103]       IW4 data #1, 74 slices, v1.2 (color), 827x1167
    BG44 [548]        IW4 data #2, 10 slices
    BG44 [265]        IW4 data #3, 6 slices
    BG44 [1118]       IW4 data #4, 7 slices

**Comment by trufanov-nok** [comment]
First of all I would say that there is no 100% good automatic segmentator. Even commercial one might replace replace a piece of illustration with a monochrome foreground symbols. It's often happens for ex. If your illustration contain a text by itself. Like a photo of a newspaper. Or if it's a map - the longitude etc. lines, country borders -  might be interpreted as a text, not an image. In this case your illustrations might be needlessly damaged, bcs some its pieces become monochrome. You can easily detect this if you Choose display mode == Foreground in DjView and find out a pieces of graphic in illustration regions.   
I'm sure 100% bulletproof auto-segmentator of scanned image is impossible to create without some strong AI. And yet it's just a fantastic.   
That's true regardless of e-book format - PDF suffer from the same problem.   
I should also notice that this problem might not affect conversion form .doc/.txm format etc. - which knows beforehand where is a text and where is an image and could benefit from this knowledge. In case of scans/photos - the convertor has no such a priori knowledge.   

Our local e-book making community do prefer DjVu to PDF bcs of filesize. The most sophisticated of them (and most productive) are aware of segmentation problem. And they avoid autosegmentation even with a commercial encoder.   
To workaround this a long time ago had been developed (by monday2000) the DjVi Imager tool. It's a program that takes your b/w multipage DjVu (bundled document), and a batch of images with the same names as the names of the pages in this document (with suffix `.sep` if I recall right). And this tool inserts these images as BG44 chunks in the corresponding pages of bundled djvu. So it inserts illustrations into already encoded multipage b/w DjVu. It uses DjVuLibre's tools for that: djvuextract, djvumake, and a slightly modified version of c44 for image encoding. c44 - is a DjVulibre's encoder used for graphics (like cjb2 - for text) to produce BG44 chunks. This is so called "method of separate scans".   

What's missing? Missing is that how they split scans to text and graphics parts. As I said, autosegmentators is no use. So they are using a semi-automatic segmentators. Semi-automatic? - These are scanned images postprocessors which try to detect illustration zones but still let user to correct them. And of course they allows user to export not only processed image, but illustration zones and all other image to separate files for further separate encoding. One of them is a ScanKromsator written by bolega for Windows. I'm not using it so don't know the details. Another is a ScanTailor by Joseph Artsimovich which is no longer maintained. But there are few forks of this project which are still developed. And one of them is mine - Scan Tailor Universal. And my work over minidjvu-mod is only to stack it with ScanTailor Universal and allow to generate djvu docs directly from it with high quality "method of separate scans" under the hood. Even for users which never hear of FG44/Sjbz chunks etc.   

Returning to `didjvu` - I don't know much about its segmenter. I just don't need segmenters. I think I tried it once and it produced comparable segmentation quality with commercial encoders - which is unacceptable for me.    
What is really interesting about `didjvu` is that it demonstrates how `FG44` chunks work. If I recall it right. Once I mentioned, that color zones are a most simple (and thus most effective in terms of the filesize) way to colorize the text. But there is another one - `FG44` chunk. It's id name hints that it's a `c44` (**c** stands for **c**oder I think, and 44 is a IW*44* wavelet encoding) encoded piece of graphic, but not for a **b**ack**g**round (BG44). It's for **F**ore**g**round - `FG44`. It's a downscaled (usually, a lot - thus compressed a lot) piece of original image corresponding to the foreground - the text part. And it's drown as a foreground using foreground text as a mask. That allows you to get colorized "automatically" text. Just detect the text part, cut the piece of original image using detected text as a mask, downscale this part of the image, encode as FG44 and it will be drown over the text layer to add some colors to it. In fact your originally colored text layer become palletizated like c*pal*djvu does, but separately from the illustration layer.  

This is a very interesting technique which is rarely demonstrated in various 3rd party DjVu tools. It allows you to have a grades of color in a single character and different color in a character that represents a several letters connected to each over due to a printing defect. Which is impossible with color zones technique of course.   

You may open the djvu you attached in DjView4 and choose Mode == Foreground and you'll see colored, not b/w text on screen. That's bcs it's colored with BG44 chunk.

So does didjvu segmentator overcome commercial? No. But you don't see this as the piece of illustration that commercial encoder would mistakenly make black may be mistakenly interpreted by `didjvu` as text too, but this text is colored! So, instead of black artifacts over illustration you'll find that some parts of your illustration are downscaled more that the rest of illustration - bcs they are in FG44, not BG44!

That's a very interesting DjVu feature that allows to keep **color** document look and feel. Even it's just a text document. I saw some 15 or 16 century polish dictionary which is encoded like that. Guess rare books worth such encoding. It had a downscaled background (which is just a downscaled parchment) and colored text over it. It Looked like it was just scanned while it's a DjVu bundled doc compressed with a djbz (e.g. similar letters are replaced with average sample) and OCR.   

It looks like a rarely needed feature (bcs our community prefer black text, white background, color/grayscale illustrations or graphics - no need in this paper color at background or shades of letters), but I see a potential here. (Still I need to admit I never worked with such docs on practice).   
Currently our librarians just shoots book pages using a planetary scanner in 300 dpi and pack them into PDF without proper processing (PDF format is a very similar to DjVu and allows a proper text encoding with dictionaries etc. - noone just uses this feature for scans. On the other side noone usually just packs its scans into DjVu without processing, bcs they have PDF for that :) So ppl think PDF and DjVu are very different while this isn't correct.). They will never use ScanTailor and tools like that to full process ebook - it's too expensive (time consuming). But with BG44 chunks and `didjvu` approach one may suggest a way to transform these PDF to a DjVus, which still looks like they are unprocessed, but the filesize 5 times smaller. Of course it could be 10 times smaller, but you can get 5 times benefit in one click, and 10 times after ears of work. And amount of such data in libraries is measured in terabytes which should be stored and accessed from web. Of course government libraries always may just ask for more funds to buy more space in a webcloud. But biggest state libraries are competent enough to prefer DjVu with FG44 to unprocessed PDFs, at least for 24/7 internet access.

**Comment by maple7-7-7** [comment]
Hi rmast,
I did experiment with didjvu a little, but was really hoping for it to have the shared dictionary function.
I have not tried the segmentation function. The use of FG44 is interesting.
[Alex] "It allows you to have grades of color in a single character . . ."
That is an interesting enhancement. Sorry I cannot help you further, rmast.

**Reply by rmast** [reply]
@maple7-7-7, I guess it will not be much work to incorporate a shared dictionary for a programmer who is interested, however there are more challenges, for example the translation from Python2.7 to Python3.8+ of didjvu and the underlying python-gamera package that's currently dropped out of Debian and Ubuntu 20.04.
@trufanov-nok, indeed this FG44 has quite a similar translation in PDF. If you process the result through DjVuToy to a PDF you'll see his PDF has a similar JPEG2000 masked by a JBIG2 and I expect it not to be much bigger that way.

**Reply by maple7-7-7** [reply]
@rmast. Thanks for the extra details about this.

**Reply by rmast** [reply]
@maple7-7-7 
> Hi rmast, I did experiment with didjvu a little, but was really hoping for it to have the shared dictionary function. 

As I read this https://github.com/jwilk/didjvu/issues/14 didjvu already must have a shared dictionary function.
cjb2 also seems to suffer the issue of too many open files.

**Reply by trufanov-nok** [reply]
@rmast 
I doubt cjb2 has "too many open files problem". It seems `didjvu` uses python2 `subprocess()` function to launch `cjb2` for each page in a loop and it's not closing handle to process' stdout, or something like that. Like this issue: https://stackoverflow.com/questions/16526783/python-subprocess-too-many-open-files     I don't know python well enough to suggest a fast fix.   
As for the shared dictionary function I surprisingly realized that they are using `minidjvu` (original) for that option: https://github.com/jwilk/didjvu/blob/master/doc/README
But I guess poring from python2 to python3 is a 1st priority task for this project..

**Reply by maple7-7-7** [reply]
@rmast
Thanks for your newest update. It was several years ago that I checked didjvu and its features.

**Comment by maple7-7-7** [comment]
Hey guys,
I know this excellent discussion of didjvu is very helpful and is in the category of General Discussion, but what do you think of placing this discussion in the Q&A category or Idea category, because this is supposed to be a General Discussion about minidjvu-mod specifically? Or I am wrong about the purpose of "General Discussion" ? I am a bit confused. Thanks.

**Reply by trufanov-nok** [reply]
I would leave it as is for now. I can create/delete or edit discussion categories if needed. The current categories are default ones that had been created by the github when you enable discussions for your project. I don't know myself what the right "purpose of "General Discussion"" is.  Discussions are quite a new github feature and I think users doesn't got accustomed to it yet. So I would rather want them to see a several discussions in one place than spread such a small number of discussions among several categories. Later, if there will be too many discussions, I could move them between groups.

**Reply by maple7-7-7** [reply]
Thanks. Sounds reasonable. We have been including Issues in the General Discussions anyway.

**Comment by rmast** [comment]
I prepared the port with Friedrich against Gamera-4, see our corresponding issue in that repository. view it on GitHub<https://github.com/jwilk/didjvu/issues/13#issuecomment-950121150>

The port runs om Mint 20.2 Python 3.8.

Verzonden vanuit Mail<https://go.microsoft.com/fwlink/?LinkId=550986> voor Windows

Van: Alexander ***@***.***>
Verzonden: zaterdag 23 oktober 2021 17:58
Aan: ***@***.***>
CC: ***@***.***>; ***@***.***>
Onderwerp: Re: [trufanov-nok/minidjvu-mod] didjvu promises to segment layers, ever tried it? (Discussion #21)


@rmast<https://github.com/rmast>
I doubt cjb2 has "too many open files problem". It seems didjvu use python2 subprocess() function to launch cjb2 for each page in a loop and it's not closing handle to process' stdout, or something like that. Like this issue: https://stackoverflow.com/questions/16526783/python-subprocess-too-many-open-files I don't know python well enough to suggest a fast fix.
As for the shared dictionary function I surprisingly realized that they are using minidjvu (original) for that option: https://github.com/jwilk/didjvu/blob/master/doc/README
But I guess poring from python2 to python3 is a 1st priority task for this project..

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/trufanov-nok/minidjvu-mod/discussions/21#discussioncomment-1525459>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAZPZ5S3YW66IT7K5GLXCQ3UILLT5ANCNFSM5GMSTUXQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by trufanov-nok** [reply]
That's a great news. Hope the project will get this update.
